cellprofiler_core/modules/metadata.py

```diff
@@ -1161,15 +1151,43 @@ def msg(url):
                     if metadata is None:
                         metadata = get_omexml_metadata(url=url)
                         filelist.add_metadata(url, metadata)
-                except:
-                    print("Metadata extraction failed for %s" % url)
+                except Exception as e:
+                    import logging
+                    logger = logging.getLogger(__name__)
+                    logger.exception(f"Metadata extraction failed for {url}", e)
                     errorcount += 1
-            if errorcount > 0:
-                wx.MessageBox(
-                    "Metadata extraction completed with %i error(s).\n"
-                    "File(s) may be in an incompatible format or "
-                    "corrupted." % errorcount
-                )
+                if errorcount > 0:
+                    msg = f"Metadata extraction completed with {errorcount} error(s).\n"
+                    "File(s) may be in an incompatible format or corrupted."
+                    if dlg is not None:
+                        import wx
+                        wx.MessageBox(msg)
+                    else:
+                        raise ValueError(msg)
+
+        is_headless = get_headless()
+        if not is_headless:
+            import wx
+            with wx.ProgressDialog(
+                    "Extracting metadata",
+                    msg(list(urls)[0]),
+                    len(urls),
+                    style=wx.PD_CAN_ABORT
+                          | wx.PD_APP_MODAL
+                          | wx.PD_ELAPSED_TIME
+                          | wx.PD_REMAINING_TIME,
+            ) as dlg:
+                def ui_callback(_dlg, i, url):
+                    if i > 0:
+                        keep_going, _ = _dlg.Update(i, msg(url))
+                        if not keep_going:
+                            return False
+                    return True
+
+                update_all_urls(dlg=dlg, ui_callback=ui_callback)
+        else:
+            update_all_urls()
```

Comment:
possibly easier to read if simplest case was first. 

more complex and maybe not worth it, maybe push the parameters into a ui_context so you can do:
```python
with get_pbar_context() as ctx:
   update_all_urls(ctx)